xdp-dump/xdpdump.c

```diff
@@ -0,0 +1,628 @@
+		snprintf(str + (i * 3), 5, "%02x ",
+			 buffer[zero_offset + i]);
+	}
+
+	if (post_skip) {
+		memset(str + (i * 3), ' ', post_skip * 3);
+		str[(i * 3) + (post_skip * 3)] = 0;
+	}
+
+	/* Print printable chars */
+	str += 16 * 3;
+	*str++ = ' ';
+
+	if (pre_skip) {
+		memset(str, ' ', pre_skip);
+		str[pre_skip] = 0;
+	}
+	for (i = pre_skip; i < 16 - post_skip; i++)
+		str[i] = isprint(buffer[zero_offset + i]) ? \
+			buffer[zero_offset + i]: '.';
+
+	str[i] = 0;
+	return 0;
+}
+
+/*****************************************************************************
+ * handle_perf_event()
+ *****************************************************************************/
+static enum bpf_perf_event_ret handle_perf_event(void *private_data,
+						 int cpu,
+						 struct perf_event_header *event)
+{
+	uint64_t ts;
+	struct perf_handler_ctx *ctx = private_data;
+	struct perf_sample_event *e = container_of(event,
+						   struct perf_sample_event,
+						   header);
+	struct perf_lost_event *lost = container_of(event,
+						    struct perf_lost_event,
+						    header);
+
+	switch(e->header.type) {
+	case PERF_RECORD_SAMPLE:
+
+		if (e->header.size < sizeof(struct perf_sample_event) ||
+		    e->size < (sizeof(struct pkt_trace_metadata) + e->metadata.cap_len))
+			return LIBBPF_PERF_EVENT_CONT;
+
+		ts = e->time + ctx->epoch_delta;
+		if (ctx->pcap_dumper) {
+			struct pcap_pkthdr h;
+
+			h.ts.tv_sec = ts / 1000000000ULL;
+			h.ts.tv_usec = ts % 1000000000ULL / 1000;
+			h.caplen = min(e->metadata.cap_len, ctx->cfg->snaplen);
+			h.len = e->metadata.pkt_len;
+			pcap_dump((u_char *) ctx->pcap_dumper, &h,
+				  e->packet);
+		} else {
+			int i;
+			char hline[SNPRINTH_MIN_BUFFER_SIZE];
+			bool fexit = e->metadata.flags & MDF_DIRECTION_FEXIT;
+
+			printf("%llu.%09lld: packet size %u bytes, "
+			       "captured %u bytes on "
+			       "if_index %u, rx queue %u, @%s\n",
+			       ts / 1000000000ULL,
+			       ts % 1000000000ULL,
+			       e->metadata.pkt_len,
+			       e->metadata.cap_len, e->metadata.ifindex,
+			       e->metadata.rx_queue,
+			       fexit ? "fexit" : "fentry");
+
+			for (i = 0; i < e->metadata.cap_len; i += 16) {
+				snprinth(hline, sizeof(hline),
+					 e->packet, e->metadata.cap_len, i);
+				printf("  %s\n", hline);
+			}
+		}
+		ctx->captured_packets++;
+		break;
+
+	case PERF_RECORD_LOST:
+		ctx->missed_events += lost->lost;
+		break;
+	}
+
+	return LIBBPF_PERF_EVENT_CONT;
+}
+
+/*****************************************************************************
+ * get_epoch_to_uptime_delta()
+ *****************************************************************************/
+static uint64_t get_epoch_to_uptime_delta(void)
+{
+	/* This function will calculate the rough delta between uptime
+	 * seconds and the epoch time. This is not a precise delta, but it's
+	 * good enough to get a general offset. The delta between packets is
+	 * still based on the timestamps from the trace infrastructure.
```

Comment:
Why is it not precise? Just because it takes a while before this is read until the first packet arrives, or?

Reply:
More the delay between the two system calls, I assume they happen at the same time.

Reply:
Yeah, sure, that's what I meant. I was just wondering if there was some other reason why this would be inaccurate, or if it was just the delay between operations; could you clarify the comment so this is clear, please?

Reply:
Will update the comment to:
`This function will calculate the rough delta between uptime seconds and the epoch time. This is not a precise delta as there is a delay between calling the two functions below (and time() being in seconds), but it's good enough to get a general offset. The delta between packets is still based on the timestamps from the trace infrastructure.`